### DISCUSSION TITLE
E-Invoice (electronic invoice pdf with xml)

### DISCUSSION THREAD

**Discussion by lwchrisS** [mpdf/mpdf#2071] [Q&A] [answered]
### Please describe the new functionality as best as you can.

Hello, will it soon be possible to generate X-invoices (e-invoices)? 

Since this will be mandatory in Germany from January 1, 2025, this would be a very helpful addition.

**Comment by finwe** [comment, marked as answer]
Hard to say as I have no knowledge of the technicalities and you didn't provide any further information.

But after a quick google search, I'd say that using [PDF/A-3 associated files + Additional XMP RDF](https://mpdf.github.io/what-else-can-i-do/pdf-a3-xmp-rdf.html) and generating ZUGFeRD metadata should be enough and is already possible.

**Comment by Kell** [comment]
When i saw https://mpdf.github.io/what-else-can-i-do/pdf-a3-xmp-rdf.html i assumed it's already possible to create PDF/A-3 files.
 But i can't figure out how to do it. I always get a PDF/A-1b.

tried different things with the given information from the mentioned link and factur-x documentations but i need a valid pdf/A-3 file.
i also found this https://github.com/mpdf/mpdf/pull/601 (from 2018) which to me looks like its possible, but how can i generate a PDF/A-3 file?
All i can find in current versions is that the code will always generate  $PDFAversion = '1-B';


just in case it helps what the current code looks like:
```php
$iccProfilePath = resource_path('icc/sRGB.icc');
if (!file_exists($iccProfilePath)) {
    die('ICC-Profil not found: ' . $iccProfilePath);
}
$mpdf = new \Mpdf\Mpdf([
    'PDFA' => true,
    'PDFAauto' => true, 
    'ICCProfilePath' => $iccProfilePath,
    'format' => 'A4',   

    'default_font' => 'opensans', 
    'fontDir' => array_merge((new \Mpdf\Config\ConfigVariables())->getDefaults()['fontDir'], [
        resource_path('fonts/opensans'), // Pfad zu den Schriftarten
    ]),

    'fontdata' => (new \Mpdf\Config\FontVariables())->getDefaults()['fontdata'] + [
            'opensans' => [
                'R' => 'OpenSans-Regular.ttf',  
                'B' => 'OpenSans-Bold.ttf',   
                'I' => 'OpenSans-Italic.ttf',  
                'BI' => 'OpenSans-BoldItalic.ttf' 
            ]
        ]
]);

$mpdf->SetAssociatedFiles([[
    'name' => 'factur-x.xml',
    'mime' => 'text/xml',
    'description' => 'some description',
    'AFRelationship' => 'Alternative',
    'path' => __DIR__ . '/../../../public/factur-x.xml'
]]);

$rdf  = '<rdf:Description rdf:about="" xmlns:zf="urn:ferd:pdfa:CrossIndustryDocument:invoice:1p0#">'."\n";
$rdf .= '  <zf:DocumentType>INVOICE</zf:DocumentType>'."\n";
$rdf .= '  <zf:DocumentFileName>factur-x.xml</zf:DocumentFileName>'."\n";
$rdf .= '  <zf:Version>1.0</zf:Version>'."\n";
$rdf .= '  <zf:ConformanceLevel>EN 16931</zf:ConformanceLevel>'."\n";
$rdf .= '</rdf:Description>'."\n";

$mpdf->SetAdditionalXmpRdf($rdf);

$mpdf->SetTitle('testpdf');

$mpdf->WriteHTML('Hello World');

$pdfFileName = "versionTest.pdf";
$pdfFilePath = "$pdfFileName";
$mpdf->Output($pdfFilePath, 'F');
```

**Reply by finwe** [reply]
You are not setting the `PDFAversion` anywhere, how would mPDF know what other version than A1-b to use? How are you testing the PDF/A-3b conformance?

**Reply by Kell** [reply]
thanks for the very fast reply. i removed it because i couldn't find that under the config: https://mpdf.github.io/reference/mpdf-variables/overview.html

i tried it again with 
`   $mpdf = new \Mpdf\Mpdf([
            'PDFA' => true,
            'PDFAversion ' => '3-B',
            'PDFAauto' => true, ...`
an i still get 1-b.
currently checking the source code and  i just added "echo $this->mpdf->PDFAversion"  in MetadataWriter.php (line 123) where this Variable is written to metadata. At this point the variable has "1-b" instead of my config "3-b"

i' currently trying to figure out whats happening in the source code.


About testing, currently i use just https://www.pdfforge.org/online/de/pdfa-validieren for validation: it shows me also :
- A file's name dictionary, as defined in PDF Reference 3.6.3, shall not contain the EmbeddedFiles key
- A file specification dictionary, as defined in PDF 3.10.2, shall not contain the EF key
These two confuse me because AFAik EF and EmbeddedFiles should be there , but unfortunately i'm have no knowledge about intern file structure of a pdf , so maybe its at wrong position.

My buddy has some special adobe which can validate  and he will check my files tomorrow.

**Reply by finwe** [reply]
You have an extra space in your config key `'PDFAversion '`, though, PDFAversion cannot be set as a part of the constructor, it must be, at least for the time being, set directly on the property `$mpdf->PDFAversion = '3-B';`

With that and your simplified example, the validator reports only

> All properties specified in XMP form shall use either the predefined schemas defined in the XMP Specification, ISO 19005-1 or this part of ISO 19005, or any extension schemas that comply with 6.6.2.3.2.

(I did not attach a correct invoice, but the `SetAdditionalXmpRdf` code is copy&pasted)

**Reply by finwe** [reply]
I was able to produce valid PDFA/3-b with this additionalXmpRdf (notice the newer Faktur-X version and therefore different namespace):

(note: not in your linked validator, but in both [this](https://www.pdflib.com/pdf-knowledge-base/xmp/free-xmp-validator/) and [this](https://avepdf.com/pdfa-validation))

```xml
<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
	<rdf:Description
	xmlns:fx="urn:ferd:pdfa:CrossIndustryDocument:invoice:2p0#" rdf:about="">
	<fx:ConformanceLevel>BASIC</fx:ConformanceLevel>
	<fx:DocumentFileName>factur-x.xml</fx:DocumentFileName>
	<fx:DocumentType>INVOICE</fx:DocumentType>
	<fx:Version>1.0</fx:Version>
	</rdf:Description>

	<rdf:Description
	xmlns:pdfaExtension="http://www.aiim.org/pdfa/ns/extension/"
	xmlns:pdfaSchema="http://www.aiim.org/pdfa/ns/schema#"
	xmlns:pdfaProperty="http://www.aiim.org/pdfa/ns/property#" rdf:about="">
	<pdfaExtension:schemas>
		<rdf:Bag>
		<rdf:li rdf:parseType="Resource">
			<pdfaSchema:schema>ZUGFeRD PDFA Extension Schema</pdfaSchema:schema>
			<pdfaSchema:namespaceURI>urn:ferd:pdfa:CrossIndustryDocument:invoice:1p0#</pdfaSchema:namespaceURI>
			<pdfaSchema:prefix>fx</pdfaSchema:prefix>
			<pdfaSchema:property>
			<rdf:Seq>
				<rdf:li rdf:parseType="Resource">
				<pdfaProperty:name>DocumentFileName</pdfaProperty:name>
				<pdfaProperty:valueType>Text</pdfaProperty:valueType>
				<pdfaProperty:category>external</pdfaProperty:category>
				<pdfaProperty:description>name of the embedded XML invoice file</pdfaProperty:description>
				</rdf:li>
				<rdf:li rdf:parseType="Resource">
				<pdfaProperty:name>DocumentType</pdfaProperty:name>
				<pdfaProperty:valueType>Text</pdfaProperty:valueType>
				<pdfaProperty:category>external</pdfaProperty:category>
				<pdfaProperty:description>INVOICE</pdfaProperty:description>
				</rdf:li>
				<rdf:li rdf:parseType="Resource">
				<pdfaProperty:name>Version</pdfaProperty:name>
				<pdfaProperty:valueType>Text</pdfaProperty:valueType>
				<pdfaProperty:category>external</pdfaProperty:category>
				<pdfaProperty:description>The actual version of the ZUGFeRD XML schema</pdfaProperty:description>
				</rdf:li>
				<rdf:li rdf:parseType="Resource">
				<pdfaProperty:name>ConformanceLevel</pdfaProperty:name>
				<pdfaProperty:valueType>Text</pdfaProperty:valueType>
				<pdfaProperty:category>external</pdfaProperty:category>
				<pdfaProperty:description>The conformance level of the embedded ZUGFeRD data</pdfaProperty:description>
				</rdf:li>
			</rdf:Seq>
			</pdfaSchema:property>
		</rdf:li>
		</rdf:Bag>
	</pdfaExtension:schemas>
	</rdf:Description>
</rdf:RDF>
```

**Reply by Kell** [reply]
Seems my last missing puzzle piece was your info about how to set the version.
I tried also with the xmp schema data from https://github.com/mpdf/mpdf/pull/601 but without setting the version properly it didn't worked.
Thanks for the help.

So with the XMP Schema from Patromo in #601 and setting properly the version i got valid PDF/A-3b in my validator and yours. PDF/A-3 validation in Adobe-3 also says its valid.
When i use your code in additionalXmpRdf i get what you get: 1-b in the validator i linked  and 3-b in yours. So i can confirm that.

Now i need to figure out proper and valid  XMP Data for to use.

thanks!

**Reply by finwe** [reply]
Great, in the next version it will also be possible to set PDFAversion in the constructor, given the key will be correct.
